### PR TITLE
feat: Redesign the gateway to be safe against race conditions

### PR DESCRIPTION
**Pull request by Bluenix2**
## Description

Previously there was all sorts of issues with the gateway and band-aid solutions for them all..

The purpose of this PR is to remove all of those for a system that should work much more safely. It adds more locks and events that make the code a bit more complicated, but it should be easier to follow.

## Changes

- Move the `BeeGees` class into a task on the class
- Make the `WebsocketClient` class an asynchronous context manager
- Update the scope and usage of `WebsocketClient` such that it now reconnects automatically
- Guard operations with a lock and event automatically
- Ensure `WebsocketClient` is safe against race conditions and handles cancellation with the asynchronous context manager

## Additional notes

This is opened as a draft as it is not yet tested (or that the race conditions that are supposed to have been solved, in fact are solved).

### CODE REVIEW

File: dis_snek/gateway.py [hunk 1]
```diff
@@ -39,177 +42,118 @@ async def rate_limit(self) -> None:
...
+        self._entered = True
+
+        self.ws = await self.state.client.http.websocket_connect(self.state.gateway_url)
+        self._closed.set()
+
+        hello = await self.receive()
+        self.heartbeat_interval = hello["d"]["heartbeat_interval"] / 1000
+
+        await asyncio.sleep(self.heartbeat_interval * random.uniform(0, 0.5))
+        self._keep_alive = asyncio.create_task(self._start_bee_gees())
```
Comment (LordOfPolls) [inline dis_snek/gateway.py]: This delays both heartbeating and indentify. Per api spec, only the heartbeat needs delaying by `heartbeat_interval * jitter`.

**Review by LordOfPolls — CHANGES_REQUESTED** [review]
Looks really good, i only had one gripe with it.
I have also tested it, and it seems to work perfectly

File: dis_snek/gateway.py (hunk 1 above)
Comment (Bluenix2) [reply dis_snek/gateway.py]: Yes I originally read it like this but whilst implementing it here I realized that surely Discord meant both the heartbeating and IDENTIFY. Heartbeating itself can't be that costly for Discord - IDENTIFY is on the other hand.

[The documentation here](https://discord.com/developers/docs/topics/gateway#identifying) implies that after starting to heartbeat (which happens after that original delay) you should IDENTIFY - hence the wait delays both the heartbeat and IDENTIFY.

I am not sure if I am right in this, but it's not much to change either way.

Comment (LordOfPolls) [reply dis_snek/gateway.py]: I don't believe your interpretation is correct. You send your receive hello, then heartbeat_interval * jitter later you send heartbeats. I don't believe identify is included in that delay.

**Review by LordOfPolls — APPROVED** [review]

**Pull request merged**